engine/pokemon/search.asm

```diff
@@ -1,81 +1,35 @@
 SpecialBeastsCheck:
-; Check if the player owns all three legendary beasts.
-; They must exist in either party or PC, and have the player's OT and ID.
+; Check if the player has caught all three legendary beasts.
 ; Return the result in hScriptVar.
 
-	ld a, RAIKOU
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	ld a, ENTEI
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	ld a, SUICUNE
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	; they exist
+	ld a, [wPlayerCaught]
+	and CAUGHT_BEASTS_MASK
+	cp CAUGHT_BEASTS_MASK
+	jr nz, SpecialMonCheckFailed
 	ld a, 1
 	ldh [hScriptVar], a
 	ret
 
 SpecialBirdsCheck:
-; Check if the player owns all three legendary birds.
-; They must exist in either party or PC, and have the player's OT and ID.
+; Check if the player has caught all three legendary birds.
 ; Return the result in hScriptVar.
 
-	ld a, ARTICUNO
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	ld a, ZAPDOS
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	ld a, MOLTRES
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	; they exist
+	ld a, [wPlayerCaught]
+	and CAUGHT_BIRDS_MASK
+	cp CAUGHT_BIRDS_MASK
+	jr nz, SpecialMonCheckFailed
 	ld a, 1
 	ldh [hScriptVar], a
 	ret
 
 SpecialDuoCheck:
-; Check if the player owns Lugia and Ho-Oh.
-; They must exist in either party or PC, and have the player's OT and ID.
-; Return the result in hScriptVar.
-
-	ld a, LUGIA
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	ld a, HO_OH
-	ldh [hScriptVar], a
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	; they exist
-	ld a, 1
-	ldh [hScriptVar], a
-	ret
-
-SpecialMonCheck:
-; Check if the player owns any monsters of the species in hScriptVar.
+; Check if the player has caught Lugia and Ho-Oh.
 ; Return the result in hScriptVar.
 
-	call CheckOwnMonAnywhere
-	jr nc, SpecialMonCheckFailed
-
-	; they exist
+	ld a, [wPlayerCaught]
+	and CAUGHT_DUO_MASK
+	cp CAUGHT_DUO_MASK
```

Comment:
https://github.com/pret/pokecrystal/wiki/Optimizing-assembly-code#compare-a-to-a-mask-after-masking-it